types/defines/rpc.d.ts

```diff
@@ -0,0 +1,115 @@
+// Namespace for RPC utility types. Unfortunately, we can't use a `module` here as these types need
+// to referenced by `Fetcher`. This is included in the "importable" version of the types which
+// strips all `module` blocks.
+declare namespace Rpc {
+  // Base type for all RPC stubs, including common memory management methods
+  interface StubBase extends Disposable {
+    dup(): this;
+  }
+
+  export type Stub<T extends object> = Provider<T> & StubBase;
+
+  // Type for method return or property on an RPC interface.
+  // For objects, use stub `Promise`s, intersected `RpcProvider`s for pipelining.
+  // Technically, we return a custom thenable here, but it quacks like a `Promise`.
+  // For primitives, just return regular `Promise`s.
+  type Result<R> = R extends object
+    ? Promise<Stub<R>> & Provider<R>
```

Comment:
Right,that's what I meant by:

> The top-level returned object from an RPC method (or property access) will have a disposer added.

Yes, we should extend `Result` to include this.

Reply:
Yeah, it's a no-op. The idea here is that you may not know if the RPC returns stubs or not -- it might even change over time, as the API evolves. So you need to be able to write code that disposes the stub properly regardless.

The only exception is if the RPC returns a primitive value, which cannot have a disposer added, but obviously doesn't ever need one.